text/0000-let-else.md

```diff
@@ -0,0 +1,473 @@
+because the compiler won't interpret it as `let PATTERN = (if y { a }) else { b };` since `()` is not an enum.
+
+This can be overcome by making a raw if-else in the expression position a compile error and instead requiring that parentheses are inserted to disambiguate:
+`let PATTERN = (if { a } else { b }) else { c };`.
+
+Rust already provides us with such a restriction, and so the expression can be restricted to be a [`ExpressionWithoutBlock`][expressions].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+let-else attempts to be as consistent as possible to similar existing syntax.
+
+Fundamentally it is treated as a `let` statement, necessitating an assignment and the trailing semicolon.
+
+Pattern matching works identically to if-let, no new "negation" pattern matching rules are introduced.
+
+Operator precedence with `&&` in made to be like if-let, requiring that a case which is an error prior to this RFC be changed to be a slightly different error.
+This is for a possible extension for let-else similar to the (yet unimplemented) if-else-chains feature, as mentioned in [future-possibilities][] with more detail.
+Specifically, while the following example is an error today, by the default `&&` operator rules it would cause problems with if-let-chains like `&&` chaining:
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean matches like this be either:
+// - Made into a compile error, or
+// - Made to be parsed internally like if-let-chains: `(let true = a) && b else { ... };`
+let true = a && b else {
+    return;
+};
+```
+
+The expression can be any [`ExpressionWithoutBlock`][expressions], in order to prevent `else {} else {}` confusion, as noted in [drawbacks][#drawbacks].
+
+The `else` must be followed by a block, as in `if {} else {}`. This else block must be diverging as the outer
+context cannot be guaranteed to continue soundly without assignment, and no alternate assignment syntax is provided.
+## Alternatives
+
+While this feature can effectively be covered by functions such `or_or`/`ok_or_else` on the `Option` and `Result` types combined with the Try operator (`?`),
+such functions do not exist automatically on custom enum types and require non-obvious and non-trivial implementation, and may not be map-able
+to `Option`/`Result`-style functions at all (especially for enums where the "success" variant is contextual and there are many variants).
+
+### `unless let ... {}` / `try let ... {}`
+
+An often proposed alternative is to add an extra keyword to the beginning of the let-else statement, to denote that it is different than a regular `let` statement.
+
+One possible benefit of adding a keyword is that it could make a possible future extension for similarity to the (yet unimplemented) [if-let-chains][] feature more straightforward.
+However, as mentioned in the [future-possibilities][] section, this is likely not necessary.
+
```

Comment:
```suggestion

One drawback of this alternative syntax: it would introduce a binding without either starting a new block containing that binding or starting with a `let`. Currently, in Rust, only a `let` statement can introduce a binding in the current block without starting a new block. This alternative syntax would potentially make it more difficult for Rust developers to scan their code for bindings, as they would need to look for both `let` and `unless let`. By contrast, a let-else statement begins with `let` and the start of a let-else statement looks exactly like a normal let binding.
```

Reply:
@Fishrock123 Fair, both `static` and `const` in fact. But this alternative would still *expand* the set of ways to start a declaration. And it's much less common to write either `static` or `const` inside a function (and they behave unusually there already).

Reply:
I updated this with a note of:

> (Note that `static` and `const` are only available outside of block scope.)

Is that incorrect?

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d3597bf80a6d65649fc38974c5414520

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4c4fb516875bba63c52683718f1df081

Reply:
I think `static` and `const` are *items*, not statements. (So in particular you can forward reference them, which you can't do with `let`.)

Reply:
Updated the note about `const` and `static`.